packages/diagram/src/layout/engine/layout.ts

```diff
@@ -0,0 +1,426 @@
+     *
+     * @param engine the engine which provides fonts
+     * @param styles styles to possibly apply to elements
+     * @param fonts fonts to use
+     * @param defaultFont the default font to use
+     */
+    constructor(
+        readonly engine: LayoutEngine,
+        readonly styles: StyleList,
+        readonly fonts: FontCollection,
+        readonly defaultFont: FontFamily
+    ) {}
+
+    /**
+     * Checks if an element matches a selector
+     * @param element the element to check
+     * @param selector the selector to check
+     * @returns true if the element matches the selector, otherwise false
+     */
+    private matchesSelector(element: LayoutElement, selector: Selector): boolean {
+        if (selector.type === SelectorType.CLASS) {
+            return element.class.has(selector.value);
+        } else if (selector.type === SelectorType.TYPE) {
+            return element.layoutConfig.type === selector.value;
+        } else if (selector.type === SelectorType.ANY) {
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    /**
+     * Checks if a style should be applied to an element
+     *
+     * @param element the element to which the style may be applied
+     * @param style the style to check
+     * @returns true if the style should be applied
+     */
+    private matchesStyle(element: LayoutElement, style: Style): boolean {
+        let currentElement: LayoutElement | undefined = element;
+        const initialElementIndex = style.selectorChain.length - 1;
+        let i = initialElementIndex;
+
+        while (currentElement) {
+            if (this.matchesSelector(currentElement, style.selectorChain[i])) {
+                i--;
+                if (i < 0) {
+                    return true;
+                }
+            } else if (i == initialElementIndex) {
+                return false;
+            }
+            currentElement = currentElement.parent;
+        }
+        return false;
+    }
+
+    /**
+     * Updates the styles of the provided layoutElement
+     *
+     * @param layoutElement the layout element where the styles should be updated
+     */
+    private applyStyles(layoutElement: LayoutElement): void {
+        const styleAttributes = layoutElement.layoutConfig.styleAttributes;
+        const matchingStyles: FullObject[] = [];
+        for (const style of this.styles.styles) {
+            if (this.matchesStyle(layoutElement, style)) {
+                matchingStyles.push(style.fields);
+            }
+        }
+        matchingStyles.push(layoutElement.element);
+        matchingStyles.reverse();
+        const styles: Record<string, any> = {};
+        for (const attributeConfig of styleAttributes) {
+            const attribute = attributeConfig.name;
+            for (const style of matchingStyles) {
+                const entry = style.getLocalFieldOrUndefined(attribute);
+                if (entry != undefined) {
```

Comment:
I've allowed a lot in the PRs until now, but nesting depth 5 is definitely too deeply nested.
Add submethods.

Reply:
fixed with 1dd0b93fa1997e740329a99042870da18b048778 and a1d8ed828b9d9e0f45263a13edef070b063da57e
(I hope 3 is good enough)